libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,709 @@
+/// Implements
+/// https://eprint.iacr.org/2018/483 and
+/// https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html
+use amcl_wrapper::{
+    constants::{GroupG1_SIZE, MODBYTES},
+    extension_field_gt::GT,
+    field_elem::FieldElement,
+    group_elem::GroupElement,
+    group_elem_g1::G1,
+    group_elem_g2::G2,
+    types_g2::GroupG2_SIZE,
+};
+use CryptoError;
+
+pub const PRIVATE_KEY_SIZE: usize = MODBYTES;
+/// This is a simple alias so the consumer can just use PrivateKey::random() to generate a new one
+/// instead of wrapping it as a private field
+pub type PrivateKey = FieldElement;
+
+macro_rules! bls_impl {
+    ($pk_size:expr, $sig_size:expr, $pk_group:ident, $sig_group:ident, $ate_2_pairing_is_one:ident, $set_pairs:ident) => {
+        pub const PUBLIC_KEY_SIZE: usize = $pk_size;
+        pub const SIGNATURE_SIZE: usize = $sig_size;
+
+        pub const MESSAGE_CONTEXT: &[u8; 20] = b"for signing messages";
+        pub const PUBLICKEY_CONTEXT: &[u8; 47] = b"for signing public keys for proof of possession";
```

Comment:
There needs to be a separator used while hashing verkeys during aggregation. [Here](https://github.com/lovesh/signature-schemes/blob/master/bls/src/multi_sig_slow.rs#L24) is an example, the `hash_verkeys` uses a domain separator

Reply:
You might want to look at the code you referenced because it doesn’t even do what are suggesting. The DS is good enough as it is

Reply:
I looked before suggesting. `hash_verkeys` is adding the separator `VERKEY_DOMAIN_PREFIX` before hashing.